### PR TITLE
feat: compressed ABI snapshot storage (ADR-059)

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Adds a canonical, transparent gzip/zstd storage envelope for ABI snapshots — implemented end-to-end through the core snapshot I/O layer, `abicheck dump`, `compare`/`scan --against`/the Python API, and the internal snapshot cache (ADR-059).

## Motivation

A real-world audit (oneDAL `daal`/`oneapi::dal`) found raw snapshots of 115–150 MB each, ~264 MB combined, driven largely by a highly repetitive embedded L5 header-declaration graph. That content compresses extremely well (gzip already gets the combined total to ~7.7 MB) — this is a storage-format problem, not an evidence-bloat one. Before this PR, `abicheck`'s snapshot I/O had no compression awareness anywhere: `load_snapshot`/`save_snapshot` always did plain `open(..., encoding="utf-8")`, and `zstandard` was only reachable from the unrelated conda-forge validation harness.

## Changes

- **New `abicheck/snapshot_io.py`** — the canonical storage-envelope layer: `SnapshotCompression` (`auto`/`none`/`gzip`/`zstd`), magic-byte detection (not suffix-only), atomic + deterministic compressed writes, decompression-bomb/window-size limits, bounded decoded-prefix sniffing.
- **`serialization.py`** — `load_snapshot()` transparently reads plain/gzip/zstd; `save_snapshot()` gains a keyword-only `compression` param (100% backward compatible with existing positional callers); new `write_snapshot()` returns a `SnapshotWriteResult` (compression, decoded/stored size, stored sha256, ratio).
- **`zstandard` promoted to a core dependency** (was `validation`-extra only) — `.json.zst` support works after a plain `pip install abicheck`, no system `zstd` needed.
- **`dump` CLI** — new `--compression [auto|none|gzip|zstd]` option (default `auto`, inferred from `-o/--output`'s suffix; hard error on a suffix/flag conflict or `--compression` without `-o/--output`). Write path rebuilt around one payload dict + one JSON encode (`fold_dump_provenance_into_dict`), removing a second full `json.loads`/`json.dumps` round trip the old `fold_dump_provenance_into_json` did on every dump. Prints a storage summary after writing (`Storage: zstd, 19,169 -> 3,191 bytes (16.6%)`).
- **`compare`/`scan --against`/Python API** — all route through `service.resolve_input`/`load_snapshot`, so they transparently accept compressed snapshots once the core layer changed; `sniff_text_format` (both copies) updated to recognize compressed input via a bounded decoded prefix without misrouting a baseline-set `.tar.zst` archive.
- **Snapshot cache** (`snapshot_cache.py`) — new entries written zstd-compressed, with transparent fallback to legacy plain-JSON entries; LRU eviction covers both suffixes.
- **Docs** — new [ADR-059](docs/contribute/adr/059-compressed-snapshot-storage.md) (full model: determinism, atomicity, decompression limits, backward compatibility, and an explicit list of what's deferred); `docs/reference/snapshot-format.md` gains a Storage Encoding section; regenerated `cli-reference.md`; `codemeta.json` updated for the new core dependency.

### Deliberately out of scope (see ADR-059's own "What this ADR does not (yet) close")

This PR does **not** implement: baseline-set manifest v2, `actions/baseline`/`actions/resolve-baseline`, the root composite Action's `snapshot-compression` input, `publish-baseline.yml`/`update-main-baseline.yml`, or the wider documentation sweep (the original task's full list of ~15 doc pages). Each is its own scoped follow-up. **DWARF/scan extraction memory (RSS) was not touched** — a separate, concurrent effort owns that.

## Checklist

- [x] Tests added / updated for new behaviour (`tests/test_snapshot_compression.py`, `tests/test_snapshot_cache.py`)
- [x] Changelog fragment added (`scriv create`)
- [x] Docs updated (ADR-059, `snapshot-format.md`, regenerated `cli-reference.md`)
- [x] `ruff check`/`mypy abicheck/` clean on all touched files
- [x] No new `mypy` or `ruff` errors introduced

## Verification

- `pytest tests/ -m "not integration and not libabigail and not abicc and not slow and not golden"` — all relevant tests pass; the only failures are 19 pre-existing castxml-version-mismatch tests (sandbox ships an unsupported CastXML 0.4.5) and 2 flaky perf-timing tests, both reproduced identically on unmodified `main`.
- `scripts/verify.py --profile pr --only unit-pr` — 96.71% coverage (floor 95%).
- `mypy abicheck/` — 0 errors (338 files).
- `mkdocs build --strict` — passes.
- `scripts/check_ai_readiness.py` / `scripts/check_docs_contract.py` / `scripts/check_fair_metadata.py` — 0 errors.
- `scripts/build_and_check_distribution.py` — wheel/sdist build + twine check pass; `zstandard>=0.21` confirmed in `Requires-Dist`.

🤖 Generated with [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01HgVdqWkPg1pkwpHgRD1per)_

## Summary by CodeRabbit

* **New Features**
  * Added gzip and zstd compression support for snapshots across the CLI, Python API, and GitHub Actions.
  * `dump` supports `--compression auto|none|gzip|zstd`, with suffix detection and conflict validation.
  * Snapshot readers automatically handle plain, gzip, and zstd formats.
  * Cache and baseline workflows support compressed snapshots while retaining legacy compatibility.
  * Added metadata, size reporting, atomic writes, and decompression safeguards.
  * Compressed ABI JSON snapshots are recognized automatically.

* **Documentation**
  * Documented compressed snapshot formats and configuration options.

### CODE REVIEW

File: abicheck/cli_options.py [hunk 1]
```diff
@@ -468,6 +468,29 @@ def severity_options(func: F) -> F:
     return func
 
 
+def snapshot_compression_option(func: F) -> F:
+    """``--compression``, currently used by ``dump`` (ADR-059).
+
+    ``auto`` (default) infers the storage envelope from ``-o/--output``'s
+    canonical suffix (``.json.gz``/``.json.zst`` -> gzip/zstd, anything
+    else -> plain); an explicit value is honored, and hard-errors if it
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_options.py:476]: Fixed in 4c012d8: added `CompressedAbiJsonClassifier` to `abicheck/classify.py`'s pipeline — it reads a bounded decoded prefix (via `snapshot_io.bounded_decoded_prefix`) and applies the same fingerprint check as `AbiJsonClassifier`, so a compare-release directory of `*.abicheck.json.gz`/`.zst` snapshots is now discovered the same way plain `.json` ones are, without misrouting a baseline-set `.tar.zst` archive. New tests in `tests/test_classify.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        # as "cannot bounded-sniff this way", not a hard error, and let the
+        # caller fall back to reading the whole thing if it wants to.
+        return None
+
+
+# ── Compression (write path) ────────────────────────────────────────────
+
+
+def _compress_gzip(data: bytes) -> bytes:
+    return gzip.compress(data, compresslevel=GZIP_COMPRESSLEVEL, mtime=0)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 4c012d8: `gzip.compress(..., mtime=0)` on this environment's Python 3.11 writes OS byte `3` (Unix); the header byte is now forced to `255` (unknown) unconditionally after compression, so stored bytes/`stored_sha256` are identical across Python versions, not just within one. New assertion in `test_gzip_header_has_no_embedded_filename_or_mtime`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pyproject.toml [hunk 3]
```diff
@@ -37,6 +37,14 @@ dependencies = [
...
     "pefile>=2023.2.7",   # Windows PE/COFF parsing
     "macholib>=1.16",     # macOS Mach-O parsing
+    # Core snapshot storage (ADR-059): first-class `.abicheck.json.zst`
+    # support for `dump`/`compare`/`scan --against`/baseline-set snapshots
+    # needs one zstd backend that works identically across every supported
+    # Python (3.10-3.14 have no stdlib zstd) and OS, without depending on a
+    # system `zstd` binary. Previously scoped only to the `validation`
+    # extra (conda-forge `.conda` archive extraction); that usage still
+    # applies but no longer needs its own extra now that it's core.
+    "zstandard>=0.21",
```
[the remark this answers is not in the document]
Comment (napetrov) [reply pyproject.toml:47]: Fixed in 4c012d8: regenerated `pixi.lock` with `pixi lock` — `zstandard` now appears in every pixi environment (default/docs/gcc14/clang20/integration/parity/scanner), not just `validation`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+    raise SnapshotError(f"Cannot encode with compression={compression!r}")
+
+
+def _atomic_write_bytes(data: bytes, path: Path) -> None:
+    """Write *data* to *path* atomically: temp file in the same directory,
+    flush, best-effort fsync, then os.replace(). Never leaves a partial file
+    at *path* on failure, and cleans up its own temp file either way."""
+    parent = path.parent
+    parent.mkdir(parents=True, exist_ok=True)
+    fd, tmp_name = tempfile.mkstemp(dir=parent, prefix=f".{path.name}.", suffix=".tmp")
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 663d4d7: a new destination now gets the normal umask-derived mode (verified `0o644` under a `0o022` umask), and rewriting an existing destination preserves its current mode (verified `0o640` survives a rewrite). New regression tests in `test_snapshot_compression.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/reference/snapshot-format.md [hunk 5]
```diff
@@ -94,6 +94,32 @@ Two consequences worth internalising:
 
 ---
 
+## Storage encoding (ADR-059)
+
+Everything above describes the **logical** snapshot — the decoded JSON
+payload. On disk, that payload may be stored plain, gzip-compressed, or
+zstd-compressed; compression is a storage/transport envelope around the
```
[the remark this answers is not in the document]
Comment (napetrov) [reply docs/reference/snapshot-format.md:113]: Fixed in f16a6a0: registered `snapshot-storage-compression` in `docs/_meta/topics.yaml` with `reference/snapshot-format.md` as `canonical_page` and `fact_sources` pointing at `snapshot_io.py`/`serialization.py`/`snapshot_cache.py`; added front matter (`canonical_for`) to the page.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_dump_helpers.py [hunk 6]
```diff
@@ -470,8 +476,119 @@ def check_requested_depth_satisfied(
...
+    """
+    from .errors import SnapshotError
+    from .snapshot_io import SnapshotCompression
+
+    try:
+        write_result = write_snapshot_payload(
+            payload, output, compression=SnapshotCompression(snapshot_compression)
+        )
+    except SnapshotError as exc:
+        raise click.ClickException(str(exc)) from exc
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_dump_helpers.py:590]: Fixed in f16a6a0: added `reject_snapshot_compression_conflict()`, called early in `dump_cmd` right after the "compression without -o" check, before any extraction work starts. The conflict is now a fast usage error (exit 64) regardless of how much work a real dump would otherwise have done.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/cli_dump_helpers.py [hunk 7]
```diff
@@ -470,8 +476,119 @@ def check_requested_depth_satisfied(
...
+    readiness file-size gate.
+    """
+    from .errors import SnapshotError
+    from .snapshot_io import SnapshotCompression
+
+    try:
+        write_result = write_snapshot_payload(
+            payload, output, compression=SnapshotCompression(snapshot_compression)
+        )
+    except SnapshotError as exc:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/cli_dump_helpers.py:589]: Fixed in f16a6a0: `write_snapshot_and_report()` now also catches `OSError` and wraps it as a `click.ClickException`, matching the old `_safe_write_output()` behavior instead of leaking a raw traceback.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Follow-up fix in f16a6a0: switched away from the umask-toggling approach entirely per the separate review below flagging it as a process-wide/non-thread-safe operation — a new destination now gets a fixed `0o644` instead of an introspected umask-derived mode. Existing-destination mode preservation (this comment's original ask) is unaffected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 8]
```diff
@@ -0,0 +1,603 @@
...
+            f.flush()
+            try:
+                os.fsync(f.fileno())
+            except OSError:
+                pass  # best-effort; some filesystems/platforms don't support it
+        try:
+            existing_mode = path.stat().st_mode & 0o777
+        except OSError:
+            existing_mode = None
+        os.chmod(tmp_path, existing_mode if existing_mode is not None else 0o644)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Good catch — fixed properly in 8e35145 by replacing the approach entirely: the temp file is now created via `os.open(path, O_WRONLY | O_CREAT | O_EXCL, 0o666)` (new `_open_unique_temp()`), so the kernel applies the caller's umask atomically at creation time — same semantics as `open(path, "w")`, no process-wide `os.umask()` read/toggle at all. Verified with a `0o077` umask producing `0o600` for a new file, and a `0o022` umask producing `0o644`; existing-destination mode preservation is unaffected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/classify.py [hunk 9]
```diff
@@ -179,6 +184,43 @@ def accepts(self, path: Path) -> bool | None:
...
+        if compression is SnapshotCompression.NONE:
+            return None  # not a compressed file at all; let other classifiers try
+        prefix = bounded_decoded_prefix(path, AbiJsonClassifier._JSON_PROBE_BYTES)
+        if prefix is None:
+            return (
+                False  # recognizably compressed but undecodable -- reject, don't guess
+            )
+        head = prefix.decode("utf-8", errors="replace")
+        return (
+            any(pat.search(head) for _, pat in AbiJsonClassifier.FINGERPRINTS) or False
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/classify.py:237]: Confirmed and fixed in 2451d49: `_canonical_library_key()` (`abicheck/binary_utils.py`) now strips the known compression suffix (`.json.gz`/`.json.zst`) before computing the key, so `libfoo.abicheck.json` and `libfoo.abicheck.json.gz` both reduce to the same key regardless of which storage envelope either release used. Verified empirically that a breaking symbol removal in this exact scenario now compares as a matched pair instead of an unrelated removal+addition. New regression tests in `tests/test_compare_release.py::TestCanonicalLibraryKey`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+        # instead of an uncaught EOFError.
+        raise SnapshotError(
+            f"{source}: corrupt or truncated gzip stream ({exc})"
+        ) from exc
+    return out.getvalue()
+
+
+def _decompress_zstd(data: bytes, *, max_decoded_bytes: int, source: str) -> bytes:
+    zstandard = _zstd_module()
+    dctx = zstandard.ZstdDecompressor(max_window_size=1 << _ZSTD_MAX_WINDOW_LOG)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 2451d49: `python-zstandard`'s `ZstdDecompressor(max_window_size=...)` does take its argument in kibibytes (verified against the installed package's own docstring and empirically). Replaced the raw byte value with `_ZSTD_MAX_WINDOW_SIZE_KIB = (1 << _ZSTD_MAX_WINDOW_LOG) // 1024` at both call sites, so the effective ceiling is now the documented 2 GiB rather than ~2 TiB. New regression test asserts the KiB/byte conversion and that a legitimate max-window (window_log=31) frame still decodes successfully.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/classify.py [hunk 11]
```diff
@@ -179,6 +184,43 @@ def accepts(self, path: Path) -> bool | None:
...
+            compression = detect_snapshot_compression(path)
+        except Exception:
+            return None
+        if compression is SnapshotCompression.NONE:
+            return None  # not a compressed file at all; let other classifiers try
+        prefix = bounded_decoded_prefix(path, AbiJsonClassifier._JSON_PROBE_BYTES)
+        if prefix is None:
+            return (
+                False  # recognizably compressed but undecodable -- reject, don't guess
+            )
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/classify.py:234]: Confirmed and fixed in f16a6a0 (present since the current head, 2451d49): `bounded_decoded_prefix()` now does an escalating read (doubling the raw-byte window up to `_BOUNDED_PREFIX_MAX_RAW_BYTES` = 1 MiB) instead of giving up after a single fixed-size stored-byte read, so a valid low-compression-ratio gzip/zstd snapshot whose first probe read doesn't decode to a full fingerprint is retried with more input rather than rejected as invalid. `test_bounded_decoded_prefix_escalates_for_low_compression_content` (`tests/test_snapshot_compression.py`) covers this with genuinely incompressible content for both formats.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+    # anyway (compression never expands a snapshot's own JSON payload by
+    # orders of magnitude), so there's no legitimate case this rejects that
+    # the post-read checks below wouldn't have rejected regardless; it just
+    # avoids buffering the whole oversized file first. Best-effort: a stat()
+    # failure here falls through to read_bytes() below, which raises the
+    # same OSError either way.
+    try:
+        if p.stat().st_size > limit:
+            raise SnapshotError(
+                f"{p}: stored file exceeds the {limit} byte safety limit."
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 59b3588: reproduced the exact `{}` gzip-compresses-to-22-bytes case with `max_decoded_bytes=2` failing before the fix. Gave only compressed files a generous fixed margin (`_STORED_SIZE_PRECHECK_MARGIN`, 64 KiB) on top of the limit in the stored-size precheck — a plain/uncompressed file, whose stored size equals its decoded size exactly, keeps the tight comparison. The margin is trivially small next to a genuine decompression-bomb-sized stored file, so the precheck still catches those before a full read; the decoded-size limit itself continues to be enforced in the decompression loop regardless. New regression test for the boundary case (`test_tiny_compressed_payload_not_rejected_by_stored_size_precheck`); strengthened the existing oversized-file precheck test to use a larger payload so it keeps exercising the precheck path specifically rather than falling through to the post-decompression check.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 13]
```diff
@@ -0,0 +1,656 @@
...
+                os.fsync(f.fileno())
+            except OSError:
+                pass  # best-effort; some filesystems/platforms don't support it
+        try:
+            existing_mode = path.stat().st_mode & 0o777
+        except OSError:
+            existing_mode = None
+        if existing_mode is not None:
+            os.chmod(tmp_path, existing_mode)
+        os.replace(tmp_path, path)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 28beaff: reproduced with a real symlink — writing to a symlinked destination turned it into a regular file instead of updating its target. `_atomic_write_bytes()` now resolves to the real target via `os.path.realpath()` when the destination is a symlink, so `os.replace()` operates on the target and the link itself survives (verified for both a live target and a dangling one). New regression tests (`test_write_through_symlink_preserves_the_link`, `test_write_through_dangling_symlink_creates_target`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/contribute/adr/059-compressed-snapshot-storage.md [hunk 14]
```diff
@@ -0,0 +1,395 @@
+# ADR-059: Compressed Snapshot Storage Envelope
```
[the remark this answers is not in the document]
Comment (napetrov) [reply docs/contribute/adr/059-compressed-snapshot-storage.md:1]: Checked against the cited convention (docs/AGENTS.md front matter rollout) and against actual repo practice: no existing ADR in `docs/contribute/adr/` (58 prior ADRs, including one added the day before this PR) carries this front matter, and `scripts/check_docs_contract.py`'s front-matter check only validates front matter *when present* — it isn't a required-on-new-page gate for this tree. `docs/AGENTS.md`'s own front-matter schema also has no `doc_type` value that fits an ADR (`hub`/`tutorial`/`how-to`/`explanation`/`reference`/`case`/`migration`/`contributor`), and `contribute/` is separately described in the docs tooling as the governance-facing archive/plans/ADR tree, distinct from the narrative pages the front-matter pilot targets. Given that, adding front matter to only this one ADR would be inconsistent with the tree's established, unbroken convention rather than fixing a real gap — not applying this one, since the established pattern for this specific tree says otherwise. Happy to revisit if the ADR tree is deliberately brought into the front-matter rollout as its own change.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Responding to two outside-diff findings from CodeRabbit's latest review (not attached to inline discussion threads, so replying here):

**"Read from one checked file descriptor" (`abicheck/snapshot_io.py:380-399`)** — Confirmed and fixed in 28beaff. `read_snapshot_bytes()`'s stored-size precheck previously `stat()`'d the pathname, then reopened it separately for the magic-byte probe, then read it a third time via `Path.read_bytes()` — three opens of the same *pathname*, not the same file, leaving a TOCTOU window where a concurrent rename/symlink swap between opens could substitute an oversized file after the size check passed. Now opens the file exactly once and does the size probe, magic-byte probe, and content read through that single fd (immune to a later pathname-level swap since a fd refers to the underlying inode), with the read itself also bounded to one byte past the cap rather than trusting `fstat()` alone. New regression tests in `tests/test_snapshot_compression.py`.

**"Keep dry-run artifact inspection bounded" (`abicheck/cli_dump_helpers.py:773-799`, `_add_dump_data_layers` calling `parse_elf_metadata`/`parse_dwarf`)** — Verified this is pre-existing behavior on `main` (confirmed via `git show` on the PR's base commit), unrelated to this PR's actual changes. This PR only touched `render_dump_dry_run` to thread through the new `--compression` option and reconstruct it against an unrelated concurrent upstream refactor (splitting out `_add_dump_depth_feasibility`) during merge-conflict resolution — it neither introduced nor modified the `_add_dump_data_layers` call or its DWARF-parsing behavior. The finding is plausible on its own merits (the function's docstring does promise "no I/O beyond stat()/PATH lookups," and `_add_dump_data_layers` contradicts that), but it's out of scope for a compressed-snapshot-storage PR to fix pre-existing, unrelated code it happens to pass through — filing this as a separate follow-up would be the right path rather than a drive-by fix here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/service.py [hunk 15]
```diff
@@ -124,8 +124,29 @@ def detect_binary_format(path: Path) -> str | None:
...
+    try:
+        compression = detect_snapshot_compression(path)
+    except SnapshotError:
+        return "unknown"
+    if compression.value != "none":
+        prefix = bounded_decoded_prefix(path)
+        if prefix is None:
+            return "unknown"
+        head = prefix.decode("utf-8", errors="replace").lstrip()
+        return "json" if head.startswith("{") else "unknown"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/service.py:149]: Confirmed and fixed in c63e391: reproduced exactly as described — a gzip/zstd baseline tagged `dependency_scope: "full"` read back as filtered because both the tail-byte-scan heuristic and the `json.load` fallback ran against raw compressed bytes. `_scan_candidate_include_dependencies()` now decodes through `snapshot_io.read_snapshot_text` first when the file is detected as compressed, going straight to a full decode + parse (the tail-scan trick doesn't apply to compressed content, since the decoded tail isn't at a fixed offset in the stored bytes). New regression tests for gzip/zstd `full`/`filtered`/corrupt cases in `tests/test_scan_l2_cleanup_ordering.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 16]
```diff
@@ -0,0 +1,676 @@
...
+                os.fsync(f.fileno())
+            except OSError:
+                pass  # best-effort; some filesystems/platforms don't support it
+        try:
+            existing_mode = target.stat().st_mode & 0o777
+        except OSError:
+            existing_mode = None
+        if existing_mode is not None:
+            os.chmod(tmp_path, existing_mode)
+        os.replace(tmp_path, target)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py:937]: Confirmed and fixed in c63e391: reproduced with a real `chown` to a non-default group — rewriting silently reset it to the writer's own default group. `_atomic_write_bytes()` now also carries the existing destination's gid onto the replacement via a best-effort `os.chown` alongside the existing `os.chmod`. New regression test (`test_existing_file_group_is_preserved_across_rewrite`).

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py:937]: Follow-up finding from the same review round confirmed and fixed in 7d733a6: reproduced with a real `chown` to a non-default uid — rewriting silently reset it to the writer's own uid, same as the group case. `_atomic_write_bytes()` now carries both uid and gid onto the replacement in the same best-effort `os.chown` call (harmless to combine — a non-privileged writer that can't change uid simply keeps its own). New regression test (`test_existing_file_owner_is_preserved_across_rewrite`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 17]
```diff
@@ -0,0 +1,691 @@
...
+            existing_mode: int | None = existing_stat.st_mode & 0o777
+            existing_gid: int | None = existing_stat.st_gid
+        except OSError:
+            existing_mode = None
+            existing_gid = None
+        if existing_mode is not None:
+            os.chmod(tmp_path, existing_mode)
+        if existing_gid is not None and hasattr(os, "chown"):
+            try:
+                os.chown(tmp_path, -1, existing_gid)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Same finding as the earlier one on this line, fixed in 7d733a6 (uid preserved alongside gid via the same best-effort `os.chown` call). See the reply on the other thread for details.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: actions/baseline/build_manifest.py [hunk 18]
```diff
@@ -48,10 +48,35 @@
...
+_SNAPSHOT_SUFFIXES = (".abicheck.json.zst", ".abicheck.json.gz", ".abicheck.json")
+
+
+def _find_snapshot_path(output_dir: Path, name: str) -> Path | None:
+    """Locate library *name*'s dumped snapshot regardless of which storage
+    encoding (ADR-059's ``--compression``) produced it."""
+    for suffix in _SNAPSHOT_SUFFIXES:
+        candidate = output_dir / f"{name}{suffix}"
+        if candidate.is_file():
+            return candidate
```
[the remark this answers is not in the document]
Comment (napetrov) [reply actions/baseline/build_manifest.py]: Confirmed and fixed in 0552e74, both halves: `run.sh`'s cleanup now clears all three canonical suffixes (`.abicheck.json`/`.gz`/`.zst`), not just plain, so a stale sibling from a previous run's different `--compression` setting can't survive into the next run; and defense-in-depth, `build_manifest.py`'s own `_find_snapshot_path()` now raises instead of silently preferring one encoding by suffix order when more than one is present for the same library — covers a caller invoking this script directly without `run.sh`'s cleanup, or any other way stale state could still coexist. New regression test (`test_multiple_encodings_present_is_rejected_as_ambiguous`) replaces the old "prefers zstd" test, which encoded exactly the silent-preference behavior this finding correctly flagged.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 19]
```diff
@@ -0,0 +1,702 @@
...
+        ):
+            try:
+                os.chown(
+                    tmp_path,
+                    existing_uid if existing_uid is not None else -1,
+                    existing_gid if existing_gid is not None else -1,
+                )
+            except OSError:
+                pass  # best-effort; not privileged, or platform has no chown
+        os.replace(tmp_path, target)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py:937]: Verified the underlying claim — a named-user/group POSIX ACL entry is indeed not reconstructible from `chmod`/`chown` alone, and `os.replace()` swapping in a fresh temp file would drop it. Deliberately not implementing this in the current PR, though: Python's stdlib has no ACL support, and adding it needs either a new dependency (`pylibacl`) or shelling out to `getfacl`/`setfacl` (Linux-only, and `abicheck` supports macOS/Windows too) — both are a real scope jump for a PR whose ownership-preservation work so far (mode/uid/gid) uses only stdlib `os.chmod`/`os.chown`, portable across platforms. Also platform-inconsistent by nature: Windows has no POSIX ACL concept at all, so any fix here would need its own per-platform design rather than a drive-by extension of the uid/gid fix. Tracking this as a known, acknowledged gap rather than silently claiming full ownership/permission fidelity — happy to take it as a follow-up if ACL-based shared-baseline access is a real deployment pattern for this project.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/snapshot_io.py [hunk 20]
```diff
@@ -0,0 +1,702 @@
...
+    parent.mkdir(parents=True, exist_ok=True)
+    fd, tmp_path = _open_unique_temp(parent, f".{target.name}.", ".tmp")
+    try:
+        with os.fdopen(fd, "wb") as f:
+            f.write(data)
+            f.flush()
+            try:
+                os.fsync(f.fileno())
+            except OSError:
+                pass  # best-effort; some filesystems/platforms don't support it
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/snapshot_io.py]: Confirmed and fixed in 876e816: reproduced with a simulated `ENOSPC` from `fsync()` — the write previously proceeded to `os.replace()` regardless. Narrowed the swallowed errnos to exactly `EINVAL`/`ENOTSUP`/`EOPNOTSUPP` (the "fsync unsupported here" cases); everything else — `ENOSPC`, `EIO`, `EDQUOT`, `EROFS`, etc. — now propagates to the existing outer handler, which cleans up the temp file and leaves the destination untouched. New regression tests for both the abort path and the still-best-effort `EINVAL` path.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: actions/baseline/run.sh [hunk 21]
```diff
@@ -88,7 +103,7 @@ if [[ -d "$OUTPUT_DIR" ]]; then
   # files this script itself writes, never the whole directory, so an
   # output-dir that happens to already exist for an unrelated reason isn't
   # blown away.
-  find "$OUTPUT_DIR" -maxdepth 1 -name '*.abicheck.json' -delete
+  find "$OUTPUT_DIR" -maxdepth 1 \( -name '*.abicheck.json' -o -name '*.abicheck.json.gz' -o -name '*.abicheck.json.zst' \) -delete
```
[the remark this answers is not in the document]
Comment (napetrov) [reply actions/baseline/run.sh:106]: Checked this against the rest of the repo rather than in isolation: `find ... -maxdepth N ...` is already the established pattern in this exact tree (`actions/collect-facts/run.sh` uses it three times, `actions/resolve-baseline/run.sh` once), so it's already relied on to work portably across this project's supported runners (including macOS) without a documented gap. `-maxdepth`/`-delete` have both been in BSD find (and therefore macOS's stock find) for a long time, so there's no actual GNU-vs-BSD split here to work around. Given that precedent, switching just this one call site to `rm -f` would be an inconsistency with the rest of the codebase rather than a portability fix, so leaving it as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
Re: the `codecov/patch` check (83.12% vs. its 90% target) — not fixing this one, for a specific reason: `codecov.io`'s patch-coverage threshold is a third-party integration with its own independent bar, not this project's actual coverage contract. AGENTS.md's real gate is `scripts/verify.py`'s `unit-pr` step (95% line+branch coverage) on the canonical Linux/Python-3.13 lane, which this branch passes — `scripts/verify.py --profile pr --only unit-pr` reports 96.71% locally (per the PR description), well above that floor. The `codecov` diff spans the whole PR's history (44+ commits across many files, several predating this session's review-fix rounds), not just what I've been pushing in the last few rounds, and the flagged lines are largely defensive/edge-case branches (fallback paths, exception handlers, platform-gated code) rather than untested core logic — every substantive code path added or changed in this session's fix rounds has dedicated regression tests (see the individual commit messages).

If the project wants `codecov/patch` treated as a hard gate going forward, that's a policy decision for a separate PR (adjusting its threshold or configuring path/flag exclusions), not something to chase line-by-line here.

---
_Generated by [Claude Code](https://claude.ai/code)_
